airflow/api_fastapi/views/ui/objects.py

```diff
@@ -0,0 +1,116 @@
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+from airflow.api_fastapi.db.common import get_session
+from airflow.api_fastapi.views.router import AirflowRouter
+from airflow.utils import timezone
+
+object_router = AirflowRouter(tags=["Object"])
+
+
+@object_router.get("/object/historical_metrics_data", include_in_schema=False)
+async def historical_metrics_data(
+    start_date: str,
+    end_date: str,
+    session: Annotated[Session, Depends(get_session)],
+) -> dict:
+    """Return cluster activity historical metrics."""
+    safe_start_date: datetime | None = _safe_parse_datetime(start_date)
+    safe_end_date: datetime | None = _safe_parse_datetime(end_date)
+
+    # DagRuns
+    dag_run_types = session.execute(
+        select(DagRun.run_type, func.count(DagRun.run_id))
+        .where(
+            DagRun.start_date >= safe_start_date,
+            func.coalesce(DagRun.end_date, timezone.utcnow()) <= safe_end_date,
+        )
+        .group_by(DagRun.run_type)
+    ).all()
+
+    dag_run_states = session.execute(
+        select(DagRun.state, func.count(DagRun.run_id))
+        .where(
+            DagRun.start_date >= safe_start_date,
+            func.coalesce(DagRun.end_date, timezone.utcnow()) <= safe_end_date,
+        )
+        .group_by(DagRun.state)
+    ).all()
+
+    # TaskInstances
+    task_instance_states = session.execute(
+        select(TaskInstance.state, func.count(TaskInstance.run_id))
+        .join(TaskInstance.dag_run)
+        .where(
+            DagRun.start_date >= safe_start_date,
+            func.coalesce(DagRun.end_date, timezone.utcnow()) <= safe_end_date,
+        )
+        .group_by(TaskInstance.state)
+    ).all()
+
+    data = {
+        "dag_run_types": {
+            **{dag_run_type.value: 0 for dag_run_type in DagRunType},
+            **dict(dag_run_types),
+        },
+        "dag_run_states": {
+            **{dag_run_state.value: 0 for dag_run_state in DagRunState},
+            **dict(dag_run_states),
+        },
+        "task_instance_states": {
+            "no_status": 0,
+            **{ti_state.value: 0 for ti_state in TaskInstanceState},
+            **{ti_state or "no_status": sum_value for ti_state, sum_value in task_instance_states},
+        },
+    }
+
+    return data
+
+
+def _safe_parse_datetime(date_to_check, allow_empty=False, strict=True) -> datetime | None:
+    """
+    Parse datetime and return error message for invalid dates.
+
+    :param date_to_check: the string value to be parsed
+    :param allow_empty: Set True to return none if empty str or None
+    :param strict: if False, it will fall back on the dateutil parser if unable to parse with pendulum
+    """
+    if allow_empty is True and not date_to_check:
+        return None
+    try:
+        return timezone.parse(date_to_check, strict=strict)
+    except (TypeError, ParserError):
+        abort(400, f"Invalid datetime: {date_to_check!r}")
```

Comment:
I think should be part of a commong `SafeDatetimeQuery` in `parameters.py`. I think there is an integration with `AfterValidatior` of pydantic to call custom validation function no query params easily.

So anybody that want's to get a safe date from query parameter can just re-use that with:
```
async def historical_metrics_data(
    start_date: SafeDateTimeQuery,
    end_date: SafeDateTimeQuery,
    session: Annotated[Session, Depends(get_session)],
) -> dict:
```

Or maybe just `DateTimeQuery` if we plan to make all of the 'date' query param safe by default.

Reply:
You can get inspiration from there https://github.com/fastapi/fastapi/discussions/10131

Reply:
Awesome one! Thanks! 

Adding this method there bugged me out too. I was going to ask what is strategy be for generalising this but I went offline quickly after pushing my last commit. You read my mind and answered my question :)

I have included the method and `AfterValidator` in `parameters.py` and added them as type definitions to the dashboard endpoint. The method is different from other Param classes but at the end, this is a parameter Annotation and included the method there as well. I am not sure if we should create a `common` or `utils` for `parameter.py` for this kind of case. How it's looking now?